File: api-server/src/main/java/com/technews/service/AuthService.java
package com.technews.service;

import com.technews.config.JwtConfig;
import com.technews.dto.request.LoginRequest;
import com.technews.dto.request.RegisterRequest;
import com.technews.dto.response.AuthResponse;
import com.technews.entity.User;
import com.technews.exception.DuplicateResourceException;
import com.technews.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtConfig jwtConfig;
    private final AuthenticationManager authenticationManager;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new DuplicateResourceException("User", "email", request.getEmail());
        }

        User user = User.builder()
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .nickname(request.getNickname())
                .build();

        user = userRepository.save(user);

        String token = jwtConfig.generateToken(
                org.springframework.security.core.userdetails.User.builder()
                        .username(user.getEmail())
                        .password(user.getPasswordHash())
                        .roles("USER")
                        .build()
        );

        return AuthResponse.builder()
                .token(token)
                .email(user.getEmail())
                .nickname(user.getNickname())
                .userId(user.getId())
                .build();
    }

    public AuthResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("User not found"));

        String token = jwtConfig.generateToken((org.springframework.security.core.userdetails.User) authentication.getPrincipal());

        return AuthResponse.builder()
                .token(token)
                .email(user.getEmail())
                .nickname(user.getNickname())
                .userId(user.getId())
                .build();
    }

    public User getCurrentUser() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String email = authentication.getName();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Current user not found"));
    }
}

File: tech-news-ios/TechNewsApp/Services/AuthService.swift
import Foundation

class AuthService {
    private let client = APIClient.shared

    func login(email: String, password: String) async throws -> AuthResponse.AuthData? {
        let request = LoginRequest(email: email, password: password)
        let response: AuthResponse = try await client.post(Constants.Endpoints.login(), body: request)

        if let authData = response.data {
            client.authToken = authData.token
            client.userEmail = authData.email
            client.userId = authData.userId
            return authData
        }
        return nil
    }

    func register(email: String, password: String, nickname: String?) async throws -> AuthResponse.AuthData? {
        let request = RegisterRequest(email: email, password: password, nickname: nickname)
        let response: AuthResponse = try await client.post(Constants.Endpoints.register(), body: request)

        if let authData = response.data {
            client.authToken = authData.token
            client.userEmail = authData.email
            client.userId = authData.userId
            return authData
        }
        return nil
    }

    func logout() {
        client.clearAuth()
    }

    func isLoggedIn() -> Bool {
        return client.authToken != nil
    }
}

// MARK: - Request Models

struct LoginRequest: Encodable {
    let email: String
    let password: String
}

struct RegisterRequest: Encodable {
    let email: String
    let password: String
    let nickname: String?
}
